### DISCUSSION TITLE
How to add custom properties on class?

### DISCUSSION THREAD

**Discussion by Kuzzy** [typegoose/typegoose#712] [Q&A] [answered]
Hi, I'm migrating my codebase from Mongoose to Typegoose and have a problem. Can you help please? I need to add few custom field to mongoose instance without `@prop` decorator (e.g. this properties should not be persisted into DB). Example:

```typescript
export class UserSchema {
  @prop({
    required: false,
    type: [String],
    enum: _.keys(constants.roles.global),
    hideJSON: true,
  })
  public roles?: Types.Array<string>

  public __dynamicRoles: string[] = []

  public get allRoles() {
    return _.union(this.__dynamicRoles, this.roles)
  }

  public addDynamicRoles(this: DocumentType<UserSchema>, ...roles) {
    console.log('>>>>>> this.__dynamicRoles', this.__dynamicRoles) // It's always `undefined`
    this.__dynamicRoles = [...this.__dynamicRoles, ...roles]
  }
}

export const User = getModelForClass(UserSchema)

// ....

const user = await User.findById('some-id')

user.addDynamicRoles(['guest', 'everyone'])

// TypeError: this.__dynamicRoles is not iterable
```

This approach works fine with Mongoose, but not with Typegoose. It's there a way to achieve such behaviour or did I've missed something? 

Thanks in advance!

**Comment by hasezoey** [comment, marked as answer]
I think you are expecting your non-prop field to have a default value, which will not work because mongoose does not "officially" support such properties and so typegoose does not translate any properties that do not have a `@prop` on them.

The reason why typegoose also does not support it is because the documents / models are not actually instances of the class you have created, classes in typegoose are basically just a combined place to have options, types and runtime schema.

**Reply by Kuzzy** [reply]
Thanks for explanation. Looks like I can still use such approach but in more dirty way without default values.
